File: components/src/commonMain/kotlin/com/zero/components/Composer.kt
package com.zero.components

import androidx.compose.runtime.Composable
import com.zero.common.component.Component
import com.zero.components.core.ButtonComponent
import com.zero.components.core.TextComponent
import com.zero.components.layouts.BoxComponent
import com.zero.components.layouts.ColumnComponent
import com.zero.components.layouts.RowComponent
import com.zero.models.core.ButtonModel
import com.zero.models.core.TextModel
import com.zero.models.layouts.BoxModel
import com.zero.models.layouts.ColumnModel
import com.zero.models.layouts.RowModel

@Composable
fun ComponentComposer(
    component: Component,
    componentModelRetriever: (String) -> Component?
) {
    when (component.state.value.componentModel) {
        // Core Components
        is ButtonModel -> ButtonComponent(component)
        is TextModel -> TextComponent(component)

        // Layout Components
        is BoxModel -> BoxComponent(component, componentModelRetriever)
        is ColumnModel -> ColumnComponent(component, componentModelRetriever)
        is RowModel -> RowComponent(component, componentModelRetriever)
    }
}

File: ios/ios/Decompose/Composer.swift
//
// Created by Gabriel Timotei on 2/21/22.
//

import SwiftUI
import common

public func ComponentComposer(id: String, componentModelRetriever: @escaping (String) -> Component?) -> AnyView {
    let component = componentModelRetriever(id)

    if let model = component?.state.value.componentModel {
        switch model {
        // Core Components
        case _ as TextModel:
            return TextComponent(component: component!).toAny()
        case _ as ButtonModel:
            return ButtonComponent(component: component!).toAny()
        // Layout Components
        case _ as BoxModel:
            return BoxComponent(component: component!, componentModelRetriever: componentModelRetriever).toAny()
        case _ as RowModel:
            return RowComponent(component: component!, componentModelRetriever: componentModelRetriever).toAny()
        case _ as ColumnModel:
            return ColumnComponent(component: component!, componentModelRetriever: componentModelRetriever).toAny()
        default:
            return EmptyView().toAny()
        }
    }

    return EmptyView().toAny()
}
